src/wasm-type.h

```diff
@@ -303,6 +299,137 @@ struct Array {
+    SignatureKind,
+    StructKind,
+    ArrayKind,
+  } kind;
+  union {
+    Signature signature;
+    Struct struct_;
+    Array array;
+  };
+  HeapType(Kind kind) : kind(kind) { assert(kind <= _last_basic_kind); }
+  HeapType(const Signature& signature)
+    : kind(SignatureKind), signature(signature) {}
+  HeapType(Signature&& signature)
+    : kind(SignatureKind), signature(std::move(signature)) {}
+  HeapType(const Struct& struct_) : kind(StructKind), struct_(struct_) {}
+  HeapType(Struct&& struct_) : kind(StructKind), struct_(std::move(struct_)) {}
+  HeapType(const Array& array) : kind(ArrayKind), array(array) {}
+  HeapType(Array&& array) : kind(ArrayKind), array(std::move(array)) {}
+  HeapType(const HeapType& other) {
+    kind = other.kind;
+    switch (kind) {
+      case FuncKind:
+      case ExternKind:
+      case AnyKind:
+      case EqKind:
+      case I31Kind:
+      case ExnKind:
+        return;
+      case SignatureKind:
+        new (&signature) auto(other.signature);
+        return;
+      case StructKind:
+        new (&struct_) auto(other.struct_);
+        return;
+      case ArrayKind:
+        new (&array) auto(other.array);
+        return;
+    }
+    WASM_UNREACHABLE("unexpected kind");
+  }
+  ~HeapType() {
+    switch (kind) {
+      case FuncKind:
+      case ExternKind:
+      case AnyKind:
+      case EqKind:
+      case I31Kind:
+      case ExnKind:
+        return;
+      case SignatureKind:
+        signature.~Signature();
+        return;
+      case StructKind:
+        struct_.~Struct();
+        return;
+      case ArrayKind:
+        array.~Array();
+        return;
+    }
+    WASM_UNREACHABLE("unexpected kind");
+  }
+  bool operator==(const HeapType& other) const {
+    if (kind != other.kind) {
+      return false;
+    }
+    switch (kind) {
+      case FuncKind:
+      case ExternKind:
+      case AnyKind:
+      case EqKind:
+      case I31Kind:
+      case ExnKind:
+        return true;
+      case SignatureKind:
+        return signature == other.signature;
+      case StructKind:
+        return struct_ == other.struct_;
+      case ArrayKind:
+        return array == other.array;
+    }
+    WASM_UNREACHABLE("unexpected kind");
+  }
+  bool operator!=(const HeapType& other) const { return !(*this == other); }
+  HeapType& operator=(const HeapType& other) {
+    if (&other != this) {
+      this->~HeapType();
+      new (this) auto(other);
+    }
+    return *this;
+  }
+  bool isSignature() const { return kind == SignatureKind; }
+  Signature getSignature() const {
+    assert(isSignature() && "Not a signature");
+    return signature;
+  }
+  bool isStruct() const { return kind == StructKind; }
+  Struct getStruct() const {
+    assert(isStruct() && "Not a struct");
+    return struct_;
+  }
+  bool isArray() const { return kind == ArrayKind; }
+  Array getArray() const {
+    assert(isArray() && "Not an array");
+    return array;
+  }
+  std::string toString() const;
+};
```

Comment:
It would be good to move the less trivial methods into the .cpp file. I'm thinking roughly anything with a switch in it. My goal there would be consistency with where the `TypeInfo` methods are placed.